engine/src/main/java/org/pentaho/di/trans/ael/websocket/TransWebSocketEngineAdapter.java

```diff
@@ -0,0 +1,532 @@
+      .filter( key -> stepHasDependency( combi, smi, key ) )
+      .flatMap( key -> opsToSteps( config.get( key ) ).stream() )
+      .collect( Collectors.toList() );
+  }
+
+  private boolean stepHasDependency( StepMetaDataCombi combi, StepMetaInterface smi, String path ) {
+    return smi.getResourceDependencies( transMeta, combi.stepMeta ).stream()
+      .flatMap( resourceReference -> resourceReference.getEntries().stream() )
+      .filter( entry -> ResourceEntry.ResourceType.ACTIONFILE.equals( entry.getResourcetype() ) )
+      .anyMatch( entry -> entry.getResource().equals( path ) );
+  }
+
+  @Override public void startThreads() throws KettleException {
+    getDaemonEndpoint().sendMessage( executionRequest );
+    //let's start the session monitor
+    startSessionMonitor();
+  }
+
+  private void startSessionMonitor() {
+    getLogChannel().logDebug( "Starting Session Monitor." );
+    sessionMonitor = Executors.newSingleThreadExecutor();
+    sessionMonitor.submit( () -> {
+      int failedTests = 0;
+      while ( !isFinished() ) {
+        try {
+          if ( failedTests > MAX_TEST_FAILED ) {
+            // Too many tests missed let's close
+            errors.incrementAndGet();
+            getLogChannel().logError(
+              "Session Monitor detected that communication with the server was lost. Finalizing execution." );
+            if ( !cancelling ) {
+              finishProcess( false );
+            }
+            // Signal for the the waitUntilFinished blocker...
+            transFinishedSignal.countDown();
+          } else {
+            TimeUnit.MILLISECONDS.sleep( SLEEP_TIME_MS );
+            getDaemonEndpoint().sessionValid();
+            if ( failedTests > 0 ) {
+              getLogChannel()
+                .logDebug( "Session Monitor - Server Communication restored." );
+            }
+            failedTests = 0;
+          }
+        } catch ( KettleException e ) {
+          failedTests++;
+          getLogChannel()
+            .logDebug(
+              "Session Monitor detected communication problem with the server. Retry (" + failedTests + "/"
+                + MAX_TEST_FAILED + ")." );
+        } catch ( InterruptedException e ) {
+          getLogChannel().logDebug( "Session Monitor was interrupted." );
+          Thread.currentThread().interrupt();
+        }
+      }
+      closeSessionMonitor();
+    } );
+  }
+
+  public void closeSessionMonitor() {
+    if ( sessionMonitor != null && !sessionMonitor.isShutdown() ) {
+      try {
+        getLogChannel().logDebug( "Shutting down the Session Monitor." );
+        sessionMonitor.shutdown();
+      } finally {
+        if ( !sessionMonitor.isTerminated() ) {
+          sessionMonitor.shutdownNow();
+        }
+        getLogChannel().logDebug( "Session Monitor shutdown." );
+      }
+    }
+  }
+
+  @Override public void waitUntilFinished() {
+    try {
+      transFinishedSignal.await();
+    } catch ( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)